csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -926,10 +803,8 @@ internal void DecDirectCount()
             {
                 // Not check for deactivation here!
                 Debug.Assert(_directCount > 0);
-                if (--_directCount == 0)
-                {
-                    System.Threading.Monitor.PulseAll(_mutex);
-                }
+                _directCount--;
+                _directCountSemaphore?.Release();
```

Comment:
Hmm, I'm pretty sure this should be:
```
if(--_directCount == 0)
{
     _directCountSemaphore?.Release();
}
```

Reply:
Right, good catch

Reply:
Actually this works because the semaphore is initialize with _directCount, but we can init it to 0 and do it the other way

Reply:
It would have to be initialized to `-_directCount` for this to work. If you initialize the semaphore with N where N > 0, the semaphore won't block for N Wait calls.